crates/ruma-client-api/src/sync/sync_events/v4.rs

```diff
@@ -0,0 +1,325 @@
+    ///
+    /// Flag which only returns rooms the user is currently invited to. If unset, both
+    /// invited and joined rooms are returned. If false, no invited rooms are returned. If
+    /// true, only invited rooms are returned.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub is_invite: Option<bool>,
+
+    /// Whether to return Rooms with tombstones, only rooms without tombstones or both.
+    ///
+    /// Flag which only returns rooms which have an `m.room.tombstone` state event. If unset,
+    /// both tombstoned and un-tombstoned rooms are returned. If false, only un-tombstoned rooms
+    /// are returned. If true, only tombstoned rooms are returned.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub is_tombstoned: Option<bool>,
+
+    /// Only list rooms of given create-types or all
+    ///
+    /// If specified, only rooms where the `m.room.create` event has a `type` matching one
+    /// of the strings in this array will be returned. If this field is unset, all rooms are
+    /// returned regardless of type. This can be used to get the initial set of spaces for an
+    /// account.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub room_types: Option<Vec<String>>,
+
+    /// Only list rooms that are not of these create-types, or all
+    ///
+    /// Same as "room_types" but inverted. This can be used to filter out spaces from the room
+    /// list.
+    #[serde(default, skip_serializing_if = "<[_]>::is_empty")]
+    pub not_room_types: Vec<String>,
+
+    /// Only list rooms matching the given string, or all
+    ///
+    /// Filter the room name. Case-insensitive partial matching e.g 'foo' matches 'abFooab'.
+    /// The term 'like' is inspired by SQL 'LIKE', and the text here is similar to '%foo%'.
+    pub room_name_like: Option<String>,
+}
+
+/// Sliding Sync Request for each list
+#[derive(Clone, Debug, Default, Serialize, Deserialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub struct SyncRequestList {
+    /// The ranges of rooms we're interested in.
+    pub ranges: Vec<(UInt, UInt)>,
+
+    /// The sort ordering applied to this list of rooms. Sticky
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub sort: Option<Vec<String>>,
+
+    /// Required state for each room returned. An array of event type and state key tuples.
+    /// Note that elements of this array are NOT sticky so they must be specified in full when they
+    /// are changed. Sticky.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub required_state: Option<Vec<(RoomEventType, String)>>,
+
+    /// The maximum number of timeline events to return per room. Sticky.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub timeline_limit: Option<UInt>,
+
+    /// Filters to apply to the list before sorting. Sticky.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub filters: Option<Raw<SyncRequestListFilters>>,
```

Comment:
I think the extensions thing is something we may want to handle the same way as [custom capabilities](https://docs.rs/ruma/latest/ruma/api/client/discovery/get_capabilities/struct.Capabilities.html#impl), rather than through `Raw` which is meant to solve potential problems with untrusted / user-generated data (`SyncRequestListFilters` is entirely server-generated, right?).

Reply:
This ties into the `extensions`-part of the spec, which isn't that clear yet, despite the fact that servers can have further extensions (and implicitly that they aren't necessarily spec'd):
> Extensions: Additional APIs which expose more data from the server e.g presence, device messages or additional sort/filter operations.

Which in this case would mean the `filters` field needs to be extensible beyond the scope of what ruma pre-defines. Isn't this what `Raw<T>` is for?

Reply:
`Raw` serves two purposes:

* it makes deserialization two-step so that user-supplied, malformed data doesn't cause deserialization of a larger structure to fail
* it wraps a part of the JSON in a way where the expected fields are easy-*ish* to access, yet other fields or the exact JSON representation aren't lost

I think in this case we don't care about the first point, or about the exact JSON representation of the expected fields, right? In that case, using a Rust structure with a catch-all field at the end would make it easier to access the expected fields which will always succeed deserialization unless there's a *server-side* bug, while also allowing access to custom fields (with a certain perf hit, but `Raw` is also not free).

Reply:
a catch all field with a `HashMap<String, JsonValue>`?

Reply:
Always `BTree`, never `Hash` in Ruma ^^

But yeah, something like that. Please make it private, and I guess if you don't want to drag this PR out for longer, we can add functionality to interact with that field separately. C.f. [capabilities](https://github.com/ruma/ruma/blob/main/crates/ruma-client-api/src/discovery/get_capabilities.rs#L68)